### PR TITLE
Api Management - 2020-12-01 stable Api-version

### PR DESCRIPTION
**Pull request by solankisamir**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js
      - [x] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Review by KacieKK — APPROVED** [review]

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/apimanagement.json [hunk 1]
```diff
@@ -0,0 +1,795 @@
...
+          "description": "Human-readable representation of the error."
+        },
+        "details": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/ErrorFieldContract"
+          },
+          "description": "The list of invalid fields send in request, in case of validation error."
+        },
+        "innerError": {
```
Comment (varunkumta) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/apimanagement.json:82]: the details property above is meant to be used to surface inner errors - ideally you should be using the ErrorResponse schema specified in the common-types/v2/types.json

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/examples/ApiManagementServiceGetDomainOwnershipIdentifier.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "parameters": {
+    "resourceGroupName": "rg1",
+    "api-version": "2020-12-01",
+    "subscriptionId": "subid"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "domainOwnershipIdentifier": ""
```
Comment (varunkumta) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/examples/ApiManagementServiceGetDomainOwnershipIdentifier.json:10]: please provide a realistic value here

**Comment by varunkumta** [comment]
>         "count": {

are you sure this works? from what I can see ARM only supports value and nextLink for collection responses.

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/definitions.json:48 in d913560. [](commit_id = d91356029585bc49b99ae0bdb68b5e224413db18, deletion_comment = False)

**Comment by varunkumta** [comment]
>         "count": {

also even if this is supported, since ARM filters collection responses based on rbac - the count might end up mismatching the number of items in the value array even after going through all pages

---
In reply to: [776419699](https://github.com/Azure/azure-rest-api-specs/pull/12904#issuecomment-776419699) [](ancestors = 776419699)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2020-12-01/definitions.json:48 in 5d5703a. [](commit_id = 5d5703acee59d6e1756f1fb84b3fea75dc3c6afe, deletion_comment = False)

**Comment by VitaliyKurokhtin** [comment]
@varunkumta, first of all, what's wrong with your comments? They're not tied to code, have these weird "refers to" and "in reply to" annotations, and I can't even reply to them.

I understand RBAC issue at hand. Right now we don't really need to know count of entities in query, the only scenario right now is when we want to validate whether certain entity exists or not, then we can make $filter=...&$top=0 request to avoid loading entity and just rely on .count >0 to understand if something is there or not. Any better way to implement this scenario?

**Comment by varunkumta** [comment]
> @varunkumta Varun Kumta FTE, first of all, what's wrong with your comments? They're not tied to code, have these weird "refers to" and "in reply to" annotations, and I can't even reply to them.
> 
> I understand RBAC issue at hand. Right now we don't really need to know count of entities in query, the only scenario right now is when we want to validate whether certain entity exists or not, then we can make $filter=...&$top=0 request to avoid loading entity and just rely on .count >0 to understand if something is there or not. Any better way to implement this scenario?

I use codeflow to review PRs - that's how it puts in comments - "Refers to" contains the file and line number information on which I made the comment, "In reply to" - is if I put in a follow-up/reply comment to another comment.

Given your requirement, and the limitation in ARM for list responses not supporting 'count' field - I would recommend adding a new post api which takes the filter query in the body and returns a corresponding result, which for you could just be "resourceExists": true/false

**Comment by VitaliyKurokhtin** [comment]
> > @varunkumta Varun Kumta FTE, first of all, what's wrong with your comments? They're not tied to code, have these weird "refers to" and "in reply to" annotations, and I can't even reply to them.
> > I understand RBAC issue at hand. Right now we don't really need to know count of entities in query, the only scenario right now is when we want to validate whether certain entity exists or not, then we can make $filter=...&$top=0 request to avoid loading entity and just rely on .count >0 to understand if something is there or not. Any better way to implement this scenario?
> 
> I use codeflow to review PRs - that's how it puts in comments - "Refers to" contains the file and line number information on which I made the comment, "In reply to" - is if I put in a follow-up/reply comment to another comment.
> 
> Given your requirement, and the limitation in ARM for list responses not supporting 'count' field - I would recommend adding a new post api which takes the filter query in the body and returns a corresponding result, which for you could just be "resourceExists": true/false

Codeflow really makes it harder to navigate feedback. One has to manually navigate to a commit and find line number in question... Anyways, suggestion makes sense, there are a couple of problems I see with it:

1. We'd have to do that for every collection.
2. This does not really conform with OData protocol with which RP API should comply. Since OData has established count annotation seems logical to use it.
3. Don't see anybody else doing that.

**Comment by varunkumta** [comment]
> > > @varunkumta
> > >             
> > >             Varun Kumta
> > >             FTE Varun Kumta FTE, first of all, what's wrong with your comments? They're not tied to code, have these weird "refers to" and "in reply to" annotations, and I can't even reply to them.
> > > I understand RBAC issue at hand. Right now we don't really need to know count of entities in query, the only scenario right now is when we want to validate whether certain entity exists or not, then we can make $filter=...&$top=0 request to avoid loading entity and just rely on .count >0 to understand if something is there or not. Any better way to implement this scenario?
> > 
> > 
> > I use codeflow to review PRs - that's how it puts in comments - "Refers to" contains the file and line number information on which I made the comment, "In reply to" - is if I put in a follow-up/reply comment to another comment.
> > Given your requirement, and the limitation in ARM for list responses not supporting 'count' field - I would recommend adding a new post api which takes the filter query in the body and returns a corresponding result, which for you could just be "resourceExists": true/false
> 
> Codeflow really makes it harder to navigate feedback. One has to manually navigate to a commit and find line number in question... Anyways, suggestion makes sense, there are a couple of problems I see with it:
> 
> 1. We'd have to do that for every collection.
> 2. This does not really conform with OData protocol with which RP API should comply. Since OData has established count annotation seems logical to use it.
> 3. Don't see anybody else doing that.

1. what kind of resource collections are these? are these tracked top level resources or nested tracked/proxy resources? If the former, this approach too will not work as arm does a fanout of list calls across regions - what happens if 1 region returns non-zero count and another zero count - ARM for sure does not have the logic of summing the counts from the 2 regions into the final response.
2. Unfortunately, even though the intention is to align with odata that has not always been achieved in practice.
3. There's always a first.

I agree I don't see a good solution either way. I would recommend posting a question on the internal stackoverflow and tagging arm-api and also coming to the next office hours

**Comment by solankisamir** [comment]
> > > > @varunkumta
> > > > ```
> > > >         Varun Kumta
> > > >         FTE Varun Kumta FTE, first of all, what's wrong with your comments? They're not tied to code, have these weird "refers to" and "in reply to" annotations, and I can't even reply to them.
> > > > ```
> > > > 
> > > > 
> > > > I understand RBAC issue at hand. Right now we don't really need to know count of entities in query, the only scenario right now is when we want to validate whether certain entity exists or not, then we can make $filter=...&$top=0 request to avoid loading entity and just rely on .count >0 to understand if something is there or not. Any better way to implement this scenario?
> > > 
> > > 
> > > I use codeflow to review PRs - that's how it puts in comments - "Refers to" contains the file and line number information on which I made the comment, "In reply to" - is if I put in a follow-up/reply comment to another comment.
> > > Given your requirement, and the limitation in ARM for list responses not supporting 'count' field - I would recommend adding a new post api which takes the filter query in the body and returns a corresponding result, which for you could just be "resourceExists": true/false
> > 
> > 
> > Codeflow really makes it harder to navigate feedback. One has to manually navigate to a commit and find line number in question... Anyways, suggestion makes sense, there are a couple of problems I see with it:
> > 
> > 1. We'd have to do that for every collection.
> > 2. This does not really conform with OData protocol with which RP API should comply. Since OData has established count annotation seems logical to use it.
> > 3. Don't see anybody else doing that.
> 
> 1. what kind of resource collections are these? are these tracked top level resources or nested tracked/proxy resources? If the former, this approach too will not work as arm does a fanout of list calls across regions - what happens if 1 region returns non-zero count and another zero count - ARM for sure does not have the logic of summing the counts from the 2 regions into the final response.

[Samir] this is only for proxy resource collection.

> 2. Unfortunately, even though the intention is to align with odata that has not always been achieved in practice.
> 3. There's always a first.
> 
> I agree I don't see a good solution either way. I would recommend posting a question on the internal stackoverflow and tagging arm-api and also coming to the next office hours

**Comment by VitaliyKurokhtin** [comment]
To expand Samir's answer. These are proxy-only resource collections that can only be requested in single region, so we should not face any issues related to ARM caching or cross region resource query. In addition, cases when customers use RBAC to limit access to resources under Microsoft.ApiManagement/service (which these collections are) are quite rare, overwhelming majority would only limit access to service itself, and not to individual resources under it. With that in mind returning count should not present any problems for majority of our users. And i'm sure we'll be able to explain the issue to a small fraction that does face discrepancy between count and actual number of entities they receive due to RBAC.

**Comment by solankisamir** [comment]
@varunkumta any update here?

**Review by zhonren — APPROVED** [review]

**Review by KedarJoshi — APPROVED** [review]

**Review by vfedonkin — APPROVED** [review]

**Review by allenjzhang — APPROVED** [review]

**Pull request merged**